frame/transaction-payment/src/lib.rs

```diff
@@ -467,6 +467,20 @@ impl<T: Trait + Send + Sync> ChargeTransactionPayment<T> where
 			Err(_) => Err(InvalidTransaction::Payment.into()),
 		}
 	}
+
+	/// Get an appropriate priority for a transaction with the given length and info.
+	///
+	/// This will try and optimise the `fee/weight` `fee/length`, whichever is consuming more of the
+	/// maximum corresponding limit.
+	///
+	/// For example, if a transaction consumed 1/4th of the block length and half of the weight, its
+	/// final priority is `fee * 2`.
+	fn get_priority(len: usize, info: &DispatchInfoOf<T::Call>, final_fee: BalanceOf<T>) -> TransactionPriority {
+		let weight_saturation = T::MaximumBlockWeight::get() / info.weight.max(1);
+		let len_saturation = T::MaximumBlockLength::get() as u64 / (len as u64).max(1);
+		let coefficient: BalanceOf<T> = weight_saturation.min(len_saturation).saturated_into::<BalanceOf<T>>();
+		final_fee.saturating_mul(coefficient).saturated_into::<TransactionPriority>()
```

Comment:
I am wondering if this operation would saturate too easily but it seems fine:

MaximumBlockWeight/BaseExtrinsicWeight ~ 10^4
MaximumBlockLength ~ 10^6
so coefficient is less than 10^6

TransactionByteFee ~ 10^10
WeightFee is at most 10^12 (MaximumBlockWeight being 10^12)
so final_fee is around 10^12 (if transaction len is not too big)

so this operation would result ~ 10^18 so it is less than 2^64 (== 10^19)

So it seems to handle all kind of transaction though not with big margin. I guess it seems fine

Reply:
yeah this could get tricky.. 

But yes I think your estimate is on the extreme end. A typical fee in a chain with 12 decimals (such as kusama) would probably be around 10^9~10^10. Given that a coefficient of even 10^6 is fine, even though that is also a very high estimate. This means that we have a transaction, which weight-wise or size-wise, is 10^6 times smaller than the block limit, i.e. a million of them can fit in a block. Do we have this in reality? I think most weights are such that we can fit a few thousand transactions in a block at most.

Reply:
Yes I agree.

My worry is if submitting a full block extrinsic saturate the priority then you can't tip beyond this. But tip is important for when you actually really need to get your transaction in.